eloquentjs/4_3_list.js

```diff
@@ -0,0 +1,119 @@
+/**
+ * Converts an array  to a list data structure
+ * @param {*} inputArray 
+ * @param {*} list 
+ * @returns 
+ */
+function arrayToList([...inputArray], list = { value: null, rest: null }) { 
+    if (!Array.isArray(inputArray)) { //this needs work
+        throw "input must be array";
+    }
+
+    if (inputArray.length === 0) {
+        return list.rest;
+    }
+
+    list.value = inputArray.pop();
+    list.rest = structuredClone(list);
+
+    return arrayToList(inputArray, list);
+}
+
+/**
+ * Converts a list data structure to an array
+ * @param {*} inputList 
+ * @param {*} arrayAccumulator 
+ * @returns 
+ */
+function listToArray(inputList, arrayAccumulator = []) {
```

Comment:
A pattern for recursion you may want to experiment with is using an inner lambda that closes over the relevant data. eg

```js
function listToArray(inputList) {
  const response = [];
  function rec(list) {
    // do your stuff in here
  }
  
  return response
}
```

It can keep the function signature cleaner, and having the structure you're accumulating in available in the closure table makes the code easier to reason about and faster a lot of the time.